### PR TITLE
Replace AddNode with SumNode

### PR DESCRIPTION
**Pull request by pet-mit**
In expressions, replace "AddNode" that takes only 2 operands, with "SumNode" that takes 0 to N operands.
This will allow shallower trees.

### CODE REVIEW

File: src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * @brief Constructs a sum node with the specified operands.
+     *
+     * @param operands The list of operands
+     */
+    template<typename... NodePtr>
+    explicit SumNode(NodePtr... operands)
+    {
+        operands_ = createVector((Node*)operands...);
+    }
```
Comment (a-zakir) [inline src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h]:
```suggestion
   
    /**
     * @brief Constructs a sum node with the specified operands.
     *
     * @param operands The list of operands
     */
    template<typename... NodePtr>
    explicit SumNode(NodePtr... operands)
    {
        
        if constexpr (sizeof...(NodePtr)) {
           operands_ = createVector((Node*)operands...);
        }
    }
```

File: src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+     */
+    std::vector<Node*> getOperands() const;
+
+    std::string name() const override
+    {
+        return "SumNode";
+    }
+
+private:
+    std::vector<Node*> operands_;
```
Comment (a-zakir) [inline src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h]:
```suggestion
    std::vector<Node*> operands_={};
```

File: src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    explicit SumNode(const std::vector<Node*>& operands);
+
+    /**
+     * @brief Retrieves the operands of the sum.
+     *
+     * @return A vector of pointers to the operands of the sum.
+     */
+    std::vector<Node*> getOperands() const;
+
+    uint getSize() const;
```
Comment (a-zakir) [inline src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h]:
```suggestion
    unsigned int getSize() const;
```

File: src/solver/expressions/nodes/SumNode.cpp [hunk 4]
```diff
@@ -18,20 +18,24 @@
...
+    operands_(operands)
+{
+}
+
+std::vector<Node*> SumNode::getOperands() const
+{
+    return operands_;
+}
+
+uint SumNode::getSize() const
```
Comment (a-zakir) [inline src/solver/expressions/nodes/SumNode.cpp]:
```suggestion
unsigned int SumNode::getSize() const
```

File: src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+     * @param operands The operands, collected in a vector
+     */
+    explicit SumNode(const std::vector<Node*>& operands);
+
+    /**
+     * @brief Retrieves the operands of the sum.
+     *
+     * @return A vector of pointers to the operands of the sum.
+     */
+    std::vector<Node*> getOperands() const;
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h]: Don't force a copy
```suggestion
    const std::vector<Node*>& getOperands() const;
    Node* operator[](std::size_t idx) const;
```

File: src/solver/expressions/iterators/pre-order.cpp [hunk 6]
```diff
@@ -18,6 +18,10 @@ std::vector<Node*> childrenLeftToRight(Node* node)
     {
         return {unary->child()};
     }
+    else if (auto* sum = dynamic_cast<SumNode*>(node))
+    {
+        return sum->getOperands();
+    }
```
Comment (flomnes) [inline src/solver/expressions/iterators/pre-order.cpp:24]: 👍🏻

File: src/solver/expressions/nodes/SumNode.cpp [hunk 7]
```diff
@@ -18,20 +18,24 @@
...
+    operands_(operands)
+{
+}
+
+std::vector<Node*> SumNode::getOperands() const
+{
+    return operands_;
+}
+
+unsigned int SumNode::getSize() const
```
Comment (flomnes) [inline src/solver/expressions/nodes/SumNode.cpp]:
```suggestion
unsigned int SumNode::size() const
```

File: src/solver/expressions/visitors/CompareVisitor.cpp [hunk 8]
```diff
@@ -55,9 +55,24 @@ static bool compareEqualOperator(const T* node, const Antares::Solver::Nodes::No
...
+bool CompareVisitor::visit(const Nodes::SumNode* node, const Nodes::Node* other)
 {
-    return compareBinaryNode(*this, node, other);
+    if (const auto* other_node = dynamic_cast<const Nodes::SumNode*>(other))
+    {
+        if (node->getSize() != other_node->getSize())
+        {
+            return false;
+        }
+        for (uint i = 0; i < node->getSize(); ++i)
```
Comment (flomnes) [inline src/solver/expressions/visitors/CompareVisitor.cpp]:
```suggestion
        for (std::size_t i = 0; i < node->size(); ++i)
```

File: src/solver/expressions/visitors/LinearityVisitor.cpp [hunk 9]
```diff
@@ -21,15 +21,22 @@
...
 {
 
-LinearStatus LinearityVisitor::visit(const Nodes::AddNode* node)
+LinearStatus LinearityVisitor::visit(const Nodes::SumNode* node)
 {
-    return dispatch(node->left()) + dispatch(node->right());
+    auto operands = node->getOperands();
+    return std::accumulate(std::begin(operands),
+                           std::end(operands),
+                           LinearStatus::CONSTANT,
```
Comment (flomnes) [inline src/solver/expressions/visitors/LinearityVisitor.cpp:37]: Is this a neutral element for + ?
LinearStatus::CONSTANT + x = x for all x

Add tests

File: src/solver/expressions/visitors/PrintVisitor.cpp [hunk 10]
```diff
@@ -19,17 +19,26 @@
...
-    // sinon erreur de compil 'fonction non trouvée'
-    return "(" + dispatch(node->left()) + "+" + dispatch(node->right()) + ")";
+std::string PrintVisitor::visit(const Nodes::SumNode* node)
+{
+    auto operands = node->getOperands();
+    if (operands.empty())
+    {
+        return "()";
+    }
+    return std::accumulate(std::begin(operands) + 1,
```
Comment (flomnes) [inline src/solver/expressions/visitors/PrintVisitor.cpp]:
```suggestion
    return std::accumulate(std::next(std::begin(operands)),
```

File: src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+std::vector<T> createVector(T first, Args... args)
+{
+    return std::vector<T>{first, args...};
+}
+
+class SumNode: public Node
+{
+public:
+    template<typename... NodePtr>
+    explicit SumNode(NodePtr... operands)
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h:44]: You need to use thr `requires` keyword here if you want to use concept `NodePtr`. Currently it's just a template parameter name.

https://en.cppreference.com/w/cpp/language/constraints

File: src/solver/expressions/visitors/CloneVisitor.cpp [hunk 12]
```diff
@@ -28,9 +28,15 @@ CloneVisitor::CloneVisitor(Registry<Nodes::Node>& registry):
...
-    return registry_.create<Nodes::AddNode>(dispatch(node->left()), dispatch(node->right()));
+Nodes::Node* CloneVisitor::visit(const Nodes::SumNode* node)
+{
+    std::vector<Nodes::Node*> clonedOperands;
+    clonedOperands.reserve(node->getSize());
+    for (auto* operand: node->getOperands())
+    {
+        clonedOperands.push_back(dispatch(operand));
+    }
+    return registry_.create<Nodes::SumNode>(clonedOperands);
```
Comment (flomnes) [inline src/solver/expressions/visitors/CloneVisitor.cpp]: You may avoid a copy with a `std::move` here, if SumNode had move constructor. 

In `SumNode`
```cpp
SumNode(std::vector<Node*>&& operands) : operands_(std::move(operands)) {}
```
Here
```cpp
// Give ownership of clonedOperands to the caller
return registry_.create<Nodes::SumNode>(std::move(clonedOperands));
```

File: src/solver/expressions/visitors/LinearityVisitor.cpp (hunk 9 above)
Comment (pet-mit) [reply src/solver/expressions/visitors/LinearityVisitor.cpp:37]: yes it is (cf. test "linear_status_plus" in test_LinearVisitor)

Comment (pet-mit) [reply src/solver/expressions/visitors/LinearityVisitor.cpp:37]: what tests would you like me to add?

Comment (flomnes) [reply src/solver/expressions/visitors/LinearityVisitor.cpp:37]: SumNode(new LiteralNode, new LiteralNode)) => CONSTANT
SumNode(new MultiplyNode(VariableNode, VariableNode)) => NON_LINEAR
etc.

File: src/tests/src/solver/expressions/test_LinearVisitor.cpp [hunk 13]
```diff
@@ -290,4 +290,30 @@ BOOST_FIXTURE_TEST_CASE(LinearityVisitor_name, Registry<Node>)
...
+    expr = create<SumNode>(new LiteralNode(41),
+                           new MultiplicationNode(new LiteralNode(5), new VariableNode("y")));
+    BOOST_CHECK_EQUAL(linearVisitor.dispatch(expr), LinearStatus::LINEAR);
+
+    expr = create<SumNode>(new MultiplicationNode(new VariableNode("x"), new VariableNode("y")));
+    BOOST_CHECK_EQUAL(linearVisitor.dispatch(expr), LinearStatus::NON_LINEAR);
+
+    expr = create<SumNode>(new LiteralNode(41),
+                           new MultiplicationNode(new LiteralNode(5), new VariableNode("y")),
+                           new MultiplicationNode(new VariableNode("x"), new VariableNode("y")));
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_LinearVisitor.cpp]: Plenty of memory leaks here, could you replace
```cpp
new Node(args...)
```
with
```cpp
create<Node>(args...)
```
?

File: src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h (hunk 11 above)
Comment (pet-mit) [reply src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h:44]: I'm not sure I understand how this should be used. Where should i use it?

**Review by flomnes — CHANGES_REQUESTED** [review]
Pending issues
- Memory leaks related to the usage of `new` in the tests
- Wrongful usage of a concept (you need to use `requires`)

File: src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h (hunk 11 above)
Comment (flomnes) [reply src/solver/expressions/include/antares/solver/expressions/nodes/SumNode.h:44]: Something like
```cpp
#include <concepts>
#include <numeric>
#include <string>

template<typename FirstArg, typename... Args>
requires(std::convertible_to<Args, FirstArg> && ...)
FirstArg sum(FirstArg first, Args... args)
{
    return first;
}

int main()
{
    sum(1);                       // OK
    sum(1, 2.3);                  // OK 2.3 is convertible to int
    sum(2, std::string("DD"));    // KO std::string is not convertible to int
    sum(1, 2, std::string("DD")); // KO std::string is not convertible to int
}
```
But then you'd need a 0 argument version : the default constructor. You can let the compiler define it.

**Review by flomnes — APPROVED** [review]

**Pull request merged**